### PR TITLE
Fixes #25233 - convert unit copy to service classes 

### PR DESCRIPTION
**Pull request by jlsherrill**

### CODE REVIEW

**Comment by jlsherrill** [comment]
[test katello]

File: app/lib/katello/util/filter_clause_generator.rb [hunk 1]
```diff
@@ -83,6 +83,10 @@ def clauses_for(list_type)
           [whitelist_all_matcher_clause] # abstract
         end
       end
+
+      def with_select_fields(clause)
+        clause #dummy method if not overridden
+      end
```
Comment (jlsherrill) [inline app/lib/katello/util/filter_clause_generator.rb]: TODO: Delete this

**Comment by jlsherrill** [comment]
Note, i've created a script for easier validations of various copy operations:  https://gist.github.com/jlsherrill/b720ba5daf706c2955994d7fa0dfacf3

You may need to comment out the alias at the top if running on a production hammer install (versus the dev setup).  Also its expecting to have access to /var/lib/pulp/, so if you are running it on a machine different from your katello server, you will need to mount /var/lib/pulp over to the system running hammer.  Either nfs or sshfs should work fine.

**Comment by jlsherrill** [comment]
[test prprocessor]

**Comment by beav** [comment]
@jlsherrill can you give more detail about the intent behind this refactor in the commit message? It has a lot of changes and may not be clear in the future.

**Comment by jlsherrill** [comment]
@beav This is the latest in a long series of changes that move pulp2 specific logic into the service class layer.  This involves pushing anything pulp-2 specific into a set of classes under app//lib/services/katello/pulp/

It was somewhat covered in our technical discussion of the pulp3 conversion on discourse.

**Comment by beav** [comment]
> @beav This is the latest in a long series of changes that move pulp2 specific logic into the service class layer. This involves pushing anything pulp-2 specific into a set of classes under app//lib/services/katello/pulp/
> 
> It was somewhat covered in our technical discussion of the pulp3 conversion on discourse.

ah, https://community.theforeman.org/t/pulp-3-migration/10474

My brain was reset by :turkey:

File: app/services/katello/pulp/repository.rb [hunk 2]
```diff
@@ -170,6 +178,16 @@ def remove_unnecessary_distributors
           smart_proxy.pulp_api.resources.repository.delete_distributor(repo.pulp_id, distributor['id']) unless found
         end
       end
+
+      def copy_units(destination_repo, units, options = {})
+        content_type = units.first.class::CONTENT_TYPE
+        uuids = units.pluck(:uuid)
+        override_config = {}
+        override_config[:recursive] = true if options[:recursive]
+        override_config[:resolve_dependencies] = true if options[:recursive]
```
Comment (sjha4) [inline app/services/katello/pulp/repository.rb]: I think you meant options[:resolve_dependencies]

**Comment by jlsherrill** [comment]
[test katello]

File: app/lib/actions/katello/repository/clone_contents.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                                :source_repo_id => source_repositories.first.id,
+                                :target_repo_id => new_repository.id).output
+
+            metadata_options[:matching_content] = match_check_output[:matching_content]
+          end
+
+          plan_action(Katello::Repository::MetadataGenerate, new_repository, metadata_options)
+        end
+
+        def lookup_rpm_ids(repo, filenames)
```
Comment (parthaa) [inline app/lib/actions/katello/repository/clone_contents.rb]: Where is this guy used?

File: app/lib/actions/katello/repository/clone_to_environment.rb [hunk 4]
```diff
@@ -19,17 +19,7 @@ def plan(repository, environment)
...
-            elsif repository.deb?
-              plan_action(Repository::CloneDebContent, repository, clone, [], false)
-            elsif repository.docker?
-              plan_action(Repository::CloneDockerContent, repository, clone, [])
-            elsif repository.ostree?
-              plan_action(Repository::CloneOstreeContent, repository, clone)
-            elsif repository.file?
-              plan_action(Repository::CloneFileContent, repository, clone)
-            end
+            plan_action(::Actions::Katello::Repository::CloneContents, [repository], clone, :copy_contents => !clone.yum?)
```
Comment (parthaa) [inline app/lib/actions/katello/repository/clone_to_environment.rb]: nice!

File: app/lib/actions/katello/repository/clone_to_version.rb [hunk 5]
```diff
@@ -14,40 +14,30 @@ def plan(repositories, content_view_version, options = {})
                                                                version: content_view_version)
 
           rpm_filenames = extract_rpm_filenames(options.fetch(:repos_units, nil), repositories.first.label)
+          fail _('Cannot publish a composite with rpm filenames') if content_view.composite? && rpm_filenames.any?
```
Comment (parthaa) [inline app/lib/actions/katello/repository/clone_to_version.rb]: is this case even possible ?

Comment (beav) [reply app/lib/actions/katello/repository/clone_to_version.rb]: We block it in the controller here: https://github.com/Katello/katello/blob/master/app/controllers/katello/api/v2/content_views_controller.rb#L88. You can export/import composites, but hammer has some logic where it handles each CVV and then export/imports the composite definition itself.

I'd be OK with leaving or removing this line. Either way, good catch! :fishing_pole_and_fish: :tropical_fish: :fish: :shark:

**Comment by parthaa** [comment]
I tested via https://gist.github.com/jlsherrill/b720ba5daf706c2955994d7fa0dfacf3  and  it came up with good results. Just need to validate non yum cases like docker/ostree.

**Comment by parthaa** [comment]
Got the following error when publishing a cv with docker repos

```
2018-12-06T23:07:05 [E|bac|e07d3] undefined local variable or method `source_repo' for #<Katello::Pulp::Repository::Docker:0x00007f7d1c3a81d0> (NameError)
 | /home/vagrant/git/katello/app/services/katello/pulp/repository/docker.rb:43:in `copy_contents'
 | /home/vagrant/git/katello/app/lib/actions/pulp/repository/copy_all_units.rb:20:in `invoke_external_task'
 | /home/vagrant/.rvm/gems/ruby-2.5.1/gems/dynflow-1.1.4/lib/dynflow/action/polling.rb:83:in `initiate_external_action'
 | /home/vagrant/.rvm/gems/ruby-2.5.1/gems/dynflow-1.1.4/lib/dynflow/action/polling.rb:18:in `run'
 | /home/vagrant/.rvm/gems/ruby-2.5.1/gems/dynflow-1.1.4/lib/dynflow/action/cancellable.rb:13:in `run'
 | /home/vagrant/git/katello/app/lib/actions/pulp/abstract_async_task.rb:45:in `run'
```

**Comment by parthaa** [comment]
Hit another error when I added a content view to a composite and tried to publish te composite
```
2018-12-06T23:32:35 [E|app|7d085] NoMethodError: undefined method `any?' for nil:NilClass
 | /home/vagrant/git/katello/app/lib/actions/katello/repository/clone_to_version.rb:17:in `plan'
 | /home/vagrant/.rvm/gems/ruby-2.5.1/gems/dynflow-1.1.4/lib/dynflow/action.rb:493:in `block (3 levels) in execute_plan'
 | /home/vagrant/.rvm/gems/ruby-2.5.1/gems/dynflow-1.1.4/lib/dynflow/middleware/stack.rb:26:in `pass'
 | /home/vagrant/.rvm/gems/ruby-2.5.1/gems/dynflow-1.1.4/lib/dynflow/middleware.rb:18:in `pass'
 | /home/vagrant/.rvm/gems/ruby-2.5.1/gems/dynflow-1.1.4/lib/dynflow/middleware.rb:35:in `plan'
```

File: app/lib/actions/katello/repository/clone_to_version.rb (hunk 5 above)
Comment (jlsherrill) [reply app/lib/actions/katello/repository/clone_to_version.rb]: yeah, a controller block is good, i think an action catch is a good idea too, as it provides another layer in the future to catch bad situations and indicates to the developer that they shouldn't have to worry about this situation later on in the action

**Comment by jlsherrill** [comment]
@parthaa updated!

**Comment by parthaa** [comment]
LGTM. APJ

**Pull request merged**